lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -389,6 +500,182 @@ class DecomposeAtenReshapeOp : public OpRewritePattern<AtenReshapeOp> {
 };
 } // namespace
 
+namespace {
+// Decompose AtenEinsumOp to AtenMmOp or AtenBmmOp
+// Step 1: split input equation to input/result tokens and find batchingDims and
+// contractingDims for future use
+// Step 2: transpose the input tensors to [batchingDims[0,1,2],
+// otherDims[0,1,2], contractingDims[0,1,2]]
+// Step 3: reshape the input tensors, the final shape should
+// be[batchingDims, otherDims, contractingDims]
+// Step 4: use AtenMmOp/AtenBmmOp to get the result, loop util we get the final
+// result
+// notice: support static shape only
+class DecomposeAtenEinsumOp : public OpRewritePattern<AtenEinsumOp> {
+ public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenEinsumOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    std::string equation;
+    if (!matchPattern(op.getEquation(), m_TorchConstantStr(equation))) {
+      return rewriter.notifyMatchFailure(op, "Unsupported value of equation");
+    }
+    SmallVector<char> resultTokens;
+    SmallVector<SmallVector<char>> inputTokens;
+    parseEquation(equation, inputTokens, resultTokens);
+
+    SmallVector<Value> inputTensors;
+    SmallVector<SmallVector<Value>> inputShapes;
+    if (!getListConstructElements(op.getTensors(), inputTensors)) {
+      return rewriter.notifyMatchFailure(
+          op, "input should comes from a PrimListConstructOp");
+    }
+
+    for (size_t i = 0; i < inputTensors.size(); i++) {
+      BaseTensorType tensorType =
+          inputTensors[i].getType().cast<BaseTensorType>();
+      if (!tensorType.hasSizes()) {
+        return rewriter.notifyMatchFailure(
+            op, "unimplemented: input tensor must have known sizes");
+      }
+      ArrayRef<int64_t> inputShape = tensorType.getSizes();
+      int64_t inputShapeSize = inputShape.size();
```

Comment:
nit: unnecessary variable. I would just do `inputShape.size()` in the `for` loop